rules/CodingStyle/Rector/Stmt/NewlineAfterStatementRector.php

```diff
@@ -98,67 +104,83 @@ public function second()
      */
     public function getNodeTypes(): array
     {
-        return [Stmt::class];
+        return [StmtsAwareInterface::class, ClassLike::class];
     }
 
     /**
-     * @param Stmt $node
-     * @return Stmt[]|null
+     * @param StmtsAwareInterface|ClassLike $node
      */
-    public function refactor(Node $node): ?array
+    public function refactor(Node $node): null|StmtsAwareInterface|ClassLike
     {
-        if (! in_array($node::class, self::STMTS_TO_HAVE_NEXT_NEWLINE, true)) {
-            return null;
-        }
-
-        $nextNode = $node->getAttribute(AttributeKey::NEXT_NODE);
-        if (! $nextNode instanceof Node) {
-            return null;
-        }
-
-        $parentCurrentNode = $node->getAttribute(AttributeKey::PARENT_NODE);
-        $parentNextNode = $nextNode->getAttribute(AttributeKey::PARENT_NODE);
+        return $this->processAddNewLine($node, false);
+    }
 
-        if ($parentNextNode !== $parentCurrentNode) {
+    private function processAddNewLine(
+        StmtsAwareInterface|ClassLike $node,
+        bool $hasChanged
+    ): null|StmtsAwareInterface|ClassLike {
+        if ($node->stmts === null) {
             return null;
         }
 
-        if ($this->shouldSkip($nextNode)) {
-            return null;
-        }
+        foreach ($node->stmts as $key => $stmt) {
+            if (! isset($node->stmts[$key + 1])) {
+                break;
+            }
 
-        $endLine = $node->getEndLine();
-        $line = $nextNode->getStartLine();
-        $rangeLine = $line - $endLine;
+            $nextStmt = $node->stmts[$key + 1];
+            if ($this->shouldSkip($nextStmt, $stmt)) {
+                continue;
+            }
 
-        if ($rangeLine > 1) {
-            /** @var Comment[]|null $comments */
-            $comments = $nextNode->getAttribute(AttributeKey::COMMENTS);
+            $endLine = $stmt->getEndLine();
+            $line = $nextStmt->getStartLine();
+            $rangeLine = $line - $endLine;
 
-            if ($this->hasNoComment($comments)) {
-                return null;
+            if ($rangeLine > 1) {
+                $rangeLine = $this->resolveRangeLineFromComment($rangeLine, $line, $endLine, $nextStmt);
             }
 
-            $phpDocInfo = $this->phpDocInfoFactory->createFromNodeOrEmpty($nextNode);
-            if ($phpDocInfo->hasChanged()) {
-                return null;
+            // skip same line or < 0 that cause infinite loop or crash
+            if ($rangeLine <= 0) {
+                continue;
             }
 
-            /** @var Comment[] $comments */
-            $line = $comments[0]->getStartLine();
-            $rangeLine = $line - $endLine;
-
             if ($rangeLine > 1) {
-                return null;
+                continue;
             }
+
+            array_splice($node->stmts, $key + 1, 0, [new Nop()]);
+
+            $hasChanged = true;
+
+            return $this->processAddNewLine($node, $hasChanged);
```

Comment:
this will slow down as repetitive call recursive will loop again on read same line, need to find a better approach.

Reply:
Handled with recursive with jump to specific keys for starter key to loop next https://github.com/rectorphp/rector-src/pull/3525/commits/97e96d51520c5f3a51ff49f44f3dce925c45106c